RecoTauTag/RecoTau/plugins/TauDiscriminationAgainstElectronDeadECAL.cc

```diff
@@ -0,0 +1,106 @@
+/** \class TauDiscriminationAgainstElectronDeadECALBase
+ *
+ * Base class for producing PFTau and PATTau discriminators
+ *
+ * Flag tau candidates reconstructed near dead ECAL channels,
+ * in order to reduce e -> tau fakes not rejected by anti-e MVA discriminator
+ *
+ * The motivation for this flag is this presentation:
+ *   https://indico.cern.ch/getFile.py/access?contribId=0&resId=0&materialId=slides&confId=177223
+ *
+ * \authors Lauri Andreas Wendland,
+ *          Christian Veelken
+ *
+ *
+ *
+ */
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "RecoTauTag/RecoTau/interface/TauDiscriminationProducerBase.h"
+#include "RecoTauTag/RecoTau/interface/AntiElectronDeadECAL.h"
+
+// helper function retrieve the correct tau type name
+template <class TauType>
+std::string getTauTypeNameString(bool capitalise) {
+  // this generic one shoudl never be called.
+  throw cms::Exception("TauDiscriminationAgainstElectronDeadECALBase")
+      << "Unsupported TauType used. You must use either reco::PFTau or pat::Tau.";
+}
+// template specialiazation to get the correct (PF/PAT)Tau type names
+template <>
+std::string getTauTypeNameString<reco::PFTau>(bool capitalise) {
+  return capitalise ? "PFRecoTau" : "pfRecoTau";
+}
+template <>
+std::string getTauTypeNameString<pat::Tau>(bool capitalise) {
+  return capitalise ? "PATTau" : "patTau";
+}
+
+template <class TauType, class TauDiscriminator>
+class TauDiscriminationAgainstElectronDeadECALBase : public TauDiscriminationProducerBase<TauType, TauDiscriminator> {
+public:
+  typedef std::vector<TauType> TauCollection;
+  typedef edm::Ref<TauCollection> TauRef;
+  explicit TauDiscriminationAgainstElectronDeadECALBase(const edm::ParameterSet& cfg)
+      : TauDiscriminationProducerBase<TauType, TauDiscriminator>::TauDiscriminationProducerBase(cfg),
+        moduleLabel_(cfg.getParameter<std::string>("@module_label")),
+        verbosity_(cfg.getParameter<int>("verbosity")),
+        antiElectronDeadECAL_(cfg) {}
+  ~TauDiscriminationAgainstElectronDeadECALBase() override {}
+
+  void beginEvent(const edm::Event& evt, const edm::EventSetup& es) override { antiElectronDeadECAL_.beginEvent(es); }
+
+  double discriminate(const TauRef& tau) const override {
+    if (verbosity_) {
+      edm::LogPrint(getTauTypeNameString<TauType>(true) + "AgainstEleDeadECAL")
```

Comment:
The base ```TauDiscriminationProducerBase.h``` has ```std::string getProducerString()``` (which should probably be moved inside the class as a static method).
It seems more practical to use the existing interface/design.

Reply:
Well, `std::string getProducerString()` and `getTauTypeNameString(bool)` do similar but not identical thing and I am not sure if they can be really unified w/o adding some logic to them. The helper function(s) can be moved to become static methods of the `TauDiscriminationProducerBase.h` class, but I do not see a big gain from it.

Reply:
From the feedback, there seems to be a good case to move ```getTauTypeNameString``` to TauDiscriminationProducerBase.h

Reply:
After some thinking I realized that making `std::string getProducerString()` a method of template `TauDiscriminationProducerBase` class requires partial specialization of the class which is not possible. Do I miss something?
Moving of  `getTauTypeNameString()` helper to `TauDiscriminationProducerBase.h` header file (or alternatively usage of another generic helper from the header) leads to errors like `invalid declaration of '::getTauTypeNameString<TauType>'`... It could be that it is something trivial, but I am not able to solve it. Help will be appreciated.

Reply:
ah, right, it's less straightforward.
Since there are just two types to care about, it can be implemented in the method, as in the following example
```c++
template <typename T, typename T1> class A {
public:
  static std::string a() {
    if (std::is_same<T, int>::value) return "int";
    if (std::is_same<T, bool>::value) return "bool";
    throw std::logic_error("BAD");
  }
```

Reply:
OK, thank you. It can be done like this.